### PR TITLE
[batch] Add k8s secret/sa cache and retry transient errors

### PR DESCRIPTION
**Pull request by jigold**
Stacked on #7825

### CODE REVIEW

**Comment by jigold** [comment]
I tested this by hand by replacing the actual function call with throwing a TimeoutError and making sure both the client and the batch-driver gave the appropriate warning in the log message and didn't proceed.

File: batch/batch/driver/main.py [hunk 1]
```diff
@@ -412,6 +413,9 @@ def validate_int(name, value, predicate, description):
     k8s_client = kube.client.CoreV1Api()
     app['k8s_client'] = k8s_client
```
Comment (danking) [inline batch/batch/driver/main.py]: this is no longer needed in the `app`, let's remove it

File: batch/batch/driver/k8s_cache.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import time
+import asyncio
+import sortedcontainers
+
+from hailtop.utils import retry_transient_errors
+
+
+class K8sCache:
```
Comment (danking) [inline batch/batch/driver/k8s_cache.py]: This is nice. It looks like the beginning of a dict an atomic, async update operation.

**Comment by jigold** [comment]
Should I close #7825 and we just merge this one? If so, I want to change the title of the PR to reflect that.

**Pull request merged**